samba 4.14.5 (new formula)

AkihiroSuda opened the pull request:
Samba (v3) was removed from Homebrew in 2016 due to technical issues (https://github.com/Homebrew/legacy-homebrew/issues/17820, https://github.com/Homebrew/homebrew-core/commit/5af93aec46c7cd43b068c9c93eb292ffd1e07fd0), however, Samba v4 seems actively supporting macOS recently, and worth including again in Homebrew.

Samba v4 is already included in [MacPorts](https://github.com/macports/macports-ports/tree/84d5040b7739311e75879d3afb4c4e1f464469d1/net/samba4).

Further discussion in https://github.com/Homebrew/discussions/discussions/1736

The formula is based on Alexander Richardson (@arichardson)'s PR #32031 .

Example usage with qemu:
```
qemu-system-x86_64 \
  -accel hvf \
  -m 4096 \
  -cdrom ubuntu-21.04-desktop-amd64.iso \
  -net nic \
  -net user,smb=/tmp/foo
```

The shared folder can be mounted with `mount.cifs //10.0.2.4/qemu /mnt` in the guest.


- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

AkihiroSuda commented:
cc @arichardson @fxcoudert

SMillerDev commented on Formula/samba.rb:
```diff
@@ -0,0 +1,54 @@
+           "--prefix=#{prefix}"
+    system "make"
+    system "make", "install"
+    # macOS has its own SMB daemon as /usr/sbin/smbd, so rename our smbd to samba-dot-org-smbd to avoid conflict.
+    # samba-dot-org-smbd is used by qemu.rb .
+    mv "#{sbin}/smbd", "#{sbin}/samba-dot-org-smbd"
```
This means the formulae should be keg-only instead of renaming IYAM

AkihiroSuda replied:
QEMU requires the binary to be renamed: https://github.com/Homebrew/homebrew-core/blob/0b1849e5735f6b1e59f5478832107c45b52e9e61/Formula/qemu.rb#L67-L72

arichardson replied:
I think it should also be fine if it's called smbd but is inside a separate directory (i.e. not linked to /usr/local/sbin). I believe the reason I used a separate name was to make the error message more obvious.

SMillerDev replied:
> QEMU requires the binary to be renamed

I think if we ship samba again we can just update the qemu formula

carlocab replied:
> I think if we ship samba again we can just update the qemu formula

Let's not do this. This breaks the install of anyone using a non-Homebrew Samba with Homebrew QEMU. Personally I think renaming one file to avoid a conflict is perfectly fine. We have a lot of formulae that do more than this to avoid becoming keg-only too.

SMillerDev replied:
Maybe we can symlink it?

AkihiroSuda replied:
macOS version of `smbd` is completely different from Samba version of `smbd` except the binary name.

So I think renaming the binary without making it keg-only is more preferable than symlinking keg-only binary.

SMillerDev commented on Formula/samba.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  test do
+    system "#{sbin}/samba-dot-org-smbd", "--version"
+    system "#{sbin}/samba-dot-org-smbd", "--help"
```
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

AkihiroSuda replied:
Added test script

carlocab commented on Formula/samba.rb:
```diff
@@ -0,0 +1,54 @@
+
+  depends_on "pkg-config" => :build
+  depends_on "gnutls"
+  depends_on "openssl"
```
```suggestion
  depends_on "openssl@1.1"
```

AkihiroSuda replied:
Sorry, fixed

carlocab commented on Formula/samba.rb:
```diff
@@ -0,0 +1,54 @@
+class Samba < Formula
+  desc "SMB/CIFS file, print, and login server for UNIX"
+  homepage "https://samba.org/"
+  license "GPL-3.0-or-later"
+  url "https://download.samba.org/pub/samba/samba-4.14.5.tar.gz"
+  sha256 "bb6ef5d2f16b85288d823578abc453d9a80514c42e5a2ea2c4e3c60dc42335c3"
```
```suggestion
  url "https://download.samba.org/pub/samba/samba-4.14.5.tar.gz"
  sha256 "bb6ef5d2f16b85288d823578abc453d9a80514c42e5a2ea2c4e3c60dc42335c3"
  license "GPL-3.0-or-later"
```

AkihiroSuda replied:
fixed

carlocab reviewed (commented):
You have some errors from `brew audit --new`. Please don't mark items in the PR template as done when you did not do them.

arichardson commented:
Is the `groups_max` patch no longer required? I don't see any changes to that function upstream. Without it samba fails if you are in more than 16 groups.

arichardson commented:
Thanks for working on this, will be great to have a binary package rather than having to build from source :) 
For the Co-authored-by, you can use the following email: `Co-authored-by: Alex Richardson <Alexander.Richardson@cl.cam.ac.uk>`

AkihiroSuda commented:
@arichardson Thanks, the `groups_patch` seems still needed. Added back the patch and added your email address.

AkihiroSuda commented:
`brew audit --new` locally succeeds but CI is failing
```
==> FAILED
samba:
  * Formulae should not require patches to build. Patches should be submitted and accepted upstream first.
Error: 1 problem in 1 formula detected
```

Is there any workaround for this?

AkihiroSuda commented:
Updated PR to use custom CFLAGS instead of applying patches, for avoiding `Formulae should not require patches to build. Patches should be submitted and accepted upstream first.` error.

Also added `caveats` to print the binary path ("/usr/local/sbin/samba-dot-org-smbd") and suggest running Samba as non-root.

AkihiroSuda commented:
CI is passing for macOS 11, but failing for macOS 10.14 and 10.15 🤔 
```
Checking for rl_completion_matches in readline                                                  : ok 
Checking for history_list in readline                                                           : ok 
Checking for SHA1_Update                                                                        : not found 
perl module Parse::Yapp::Driver 1.05                                                            : not found 
perl module "Parse::Yapp::Driver" not found
```

SMillerDev commented on Formula/samba.rb:
```diff
@@ -0,0 +1,159 @@
+      interfaces=127.0.0.1
+      bind interfaces only=yes
+      pid directory=${state_dir}
+      lock directory=${state_dir}
+      state directory=${state_dir}
+      cache directory=${state_dir}
+      ncalrpc dir=${state_dir}/ncalrpc
+      log file=${state_dir}/log.smbd
+      smb passwd file=${state_dir}/smbpasswd
+      security = user
+      map to guest = Bad User
+      load printers = no
+      printing = bsd
+      disable spoolss = yes
+      usershare max shares = 0
+      [test]
+      path=${data_dir}
+      read only=no
+      guest ok=yes
+      force user=$(whoami)
+      EOF
+
+      echo "=== Starting ${SMBD} on 127.0.0.1:${PORT} ==="
+      "$SMBD" -F --configfile="${smb_conf}" --port="${PORT}" &
+      smbd_pid=$!
+
+      echo "=== Sleeping until the daemon starts up ==="
+      sleep 3
+
+      mkdir -p "${tmp_dir}/got"
+      echo "=== Downloading \"//127.0.0.1/test/hello\" ==="
+      smbclient -p "${PORT}" -N //127.0.0.1/test -c "get hello ${tmp_dir}/got/hello"
+
+      echo "=== Stopping ${SMBD} (PID ${smbd_pid}) ==="
+      kill "$smbd_pid"
+      smbd_pid=
+      echo "=== Test is passing ==="
+    EOS
+
+    system "bash", "#{testpath}/samba-smoke-test.sh"
```
```suggestion
    port = free_port
    
    mkdir_p "samba/state"
    mkdir_p "samba/data"
    (testpath/"samba/data/hello").write "hello"

    # mimic smb.conf generated by qemu
    # https://github.com/qemu/qemu/blob/v6.0.0/net/slirp.c#L862
    (testpath/"smb.conf") <<~EOS
      [global]
      private dir=#{testpath}/samba/state
      interfaces=127.0.0.1
      bind interfaces only=yes
      pid directory=#{testpath}/samba/state
      lock directory=#{testpath}/samba/state
      state directory=#{testpath}/samba/state
      cache directory=#{testpath}/samba/state
      ncalrpc dir=#{testpath}/samba/state/ncalrpc
      log file=#{testpath}/samba/state/log.smbd
      smb passwd file=#{testpath}/samba/state/smbpasswd
      security = user
      map to guest = Bad User
      load printers = no
      printing = bsd
      disable spoolss = yes
      usershare max shares = 0
      [test]
      path=#{testpath}/samba/data
      read only=no
      guest ok=yes
      force user=#{ENV["USER"]}
    EOS

    on_macos do
      pid = fork { exec sbin/"samba-dot-org-smbd", "-F", "--configfile=smb.conf", "--port=#{port}" }
    end

    on_linux do
      pid = fork { exec sbin/"smbd", "-F", "--configfile=smb.conf", "--port=#{port}" }
    end
     
    sleep 3

    mkdir_p "got"
    system "smbclient", "-p", "#{port}", "-N", "//127.0.0.1/test", "-c", "get hello #{testpath}/got/hello"
```

AkihiroSuda replied:
Thanks, applied

SMillerDev commented on Formula/samba.rb:
```diff
@@ -0,0 +1,133 @@
+    mkdir_p "got"
+    system "smbclient", "-p", port.to_s, "-N", "//127.0.0.1/test", "-c", "get hello #{testpath}/got/hello"
+    assert_equal "hello", (testpath/"got/hello").read
+  ensure
+    Process.kill("TERM", pid) if pid
+  end
```
This isn't needed, brew will do this already.

AkihiroSuda replied:
Thanks, deleted

SMillerDev commented on Formula/samba.rb:
```diff
@@ -0,0 +1,133 @@
+    EOS
+
+    pid = 0
+    on_macos do
+      pid = fork { exec sbin/"samba-dot-org-smbd", "-F", "--configfile=smb.conf", "--port=#{port}" }
+    end
+    on_linux do
+      pid = fork { exec sbin/"smbd", "-F", "--configfile=smb.conf", "--port=#{port}" }
+    end
```
I believe this needs 
```suggestion
    on_macos do
      pid = fork { exec sbin/"samba-dot-org-smbd", "-S", "-F", "--configfile=smb.conf", "--port=#{port}" }
    end
    on_linux do
      pid = fork { exec sbin/"smbd", "-S", "-F", "--configfile=smb.conf", "--port=#{port}" }
    end
```

AkihiroSuda replied:
Added

carlocab commented on Formula/samba.rb:
```diff
@@ -0,0 +1,123 @@
+    if MacOS.version < :big_sur
+      ENV.prepend_create_path "PERL5LIB", libexec/"lib/perl5"
+      ENV.prepend_path "PATH", libexec/"bin"
+      resource("Parse::Yapp").stage do
```
Is this resource needed at runtime? How does `samba` know where to find it at runtime, if it's needed then?

arichardson replied:
It's been a long time since I looked at samba but I'm pretty sure it's only needed for the build and not required at run time.

carlocab replied:
If it's only needed at build time, install it into `buildpath` rather than `libexec`.

AkihiroSuda replied:
fixed

AkihiroSuda commented on Formula/samba.rb:
```diff
@@ -0,0 +1,123 @@
+    EOS
+
+    port = free_port
+    fork { exec smbd, "-S", "-F", "--configfile=smb.conf", "--port=#{port}", "--debuglevel=10" }
```
`brew test` passes locally on my macOS 11 laptop but CI failing with `Server exit (EOF on stdin)` 

Maybe we need to supply custom stdin FD when TTY is detached?

carlocab replied:
Possibly. You can use `spawn` to easily redirect stdin. See `neovim-remote` for an example.

AkihiroSuda replied:
Added `spawn`, thanks

carlocab commented on Formula/samba.rb:
```diff
@@ -0,0 +1,123 @@
+    end
+    # CFLAGS is for avoiding hitting https://bugzilla.samba.org/show_bug.cgi?id=14680 .
+    # Remove this CFLAGS when the patch ( https://attachments.samba.org/attachment.cgi?id=16579 ) gets merged.
+    ENV.append "CFLAGS", "-include #{buildpath}/lib/util/debug.h"
```
This feels like a pretty heavy-handed way of getting around the no-patches requirement.
